spec/immutable-data-structures.html

```diff
@@ -506,6 +506,53 @@ <h1>Tuple.prototype.filter ( _callbackfn_ [ , _thisArg_ ] )</h1>
         <h1>Tuple.prototype.find</h1>
         <p>The initial value of *Tuple.prototype.find* is  %Array.prototype.find%.</p>
       </emu-clause>
+      <emu-clause id="sec-tuple.prototype.flat">
+        <h1>Tuple.prototype.flat ( [ _depth_ ] )</h1>
+        <p>When the `flat` method is called with zero or one arguments, the following steps are taken:</p>
+        <emu-alg>
+          1. Let _T_ be ? thisTupleValue(*this* value).
+          1. Let _list_ be _T_.[[Sequence]].
+          1. Let _depthNum_ be 1.
+          1. If _depth_ is not *undefined*, then
+            1. Set _depthNum_ to ? ToInteger(_depth_).
+          1. Let _flat_ be a new empty List.
+          1. Perform ? FlattenIntoList(_flat_, _list_, _depthNum_).
+          1. Return a new Tuple value whose [[Sequence]] is _flat_.
+        </emu-alg>
+
+        <emu-clause id="sec-flattenintolist" aoid="FlattenIntoList">
+          <h1>FlattenIntoList ( _target_, _source_, _depth_ [ , _mapperFunction_, _thisArg_ ] )</h1>
+          <p>The abstract operation FlattenIntoList takes arguments _target_, _source_, and _depth_ and optional arguments _mapperFunction_ and _thisArg_. It performs the following steps when called:</p>
+          <emu-alg>
+            1. Assert: _target_ is a List.
+            1. Assert: _source_ is a List.
+            1. Assert: ! IsInteger(_depth_) is *true*, or _depth_ is either *+&infin;* or *-&infin;*.
+            1. Assert: If _mapperFunction_ is present, then ! IsCallable(_mapperFunction_) is *true*, _thisArg_ is present, and _depth_ is *1*.
+            1. For each _element_ of _source_,
+              1. If _mapperFunction_ is present, then
+                1. Set _element_ to ? Call(_mapperFunction_, _thisArg_, &laquo; _element_, _sourceIndex_, _source_ &raquo;).
+                1. If Type(_element_) is Object, throw a *TypeError* exception.
```

Comment:
Hmm, I wonder if we should permit objects here as long as it all flattens into non-objects. `Tuple.prototype.concat` is flexible in this way. But on the other hand, `Array.prototype.flat{,Map}` just check IsArray, so checking for a Tuple primitive feels to me like the equivalent. So, these semantics seem good to me for a first pass.

Reply:
why wouldn't this also flatten boxed Tuple objects?

Reply:
I think you are talking about two different things.

@littledan Do you mean something like `tuple.flatMap(x => [1, 2])`? It makes sense to me.

@ljharb The other tuple methods don't unbox. I don't have a strong opinion about this, other than that all the methods should be consistent.

Reply:
I would expect them all to unbox - https://tc39.es/ecma262/#sec-tonumber for example calls ToPrimitive on objects, and ToNumber is called on anything that expects numbers.

Similarly, I'd expect everything that expects a Tuple to use a ToTuple operation that calls ToPrimitive.

Reply:
All the Tuple methods unbox the receiver. In other cases like JSON.stringify, we unbox wrappers (since they are giving special handling to Tuples). However, literal syntax and Record.fromEntries, Tuple.of, etc do not unbox (since they are checking for primitiveness in general). I don't think there is a strong precedent one way or the other, but unboxing seems reasonable to me in this case.